File: android/src/main/java/com/mrousavy/camera/frameprocessor/FrameProcessorPluginRegistry.java
package com.mrousavy.camera.frameprocessor;

import androidx.annotation.Keep;
import androidx.annotation.Nullable;
import com.facebook.proguard.annotations.DoNotStrip;
import java.util.Map;
import java.util.HashMap;

@DoNotStrip
@Keep
public class FrameProcessorPluginRegistry {
    private static final Map<String, PluginInitializer> Plugins = new HashMap<>();

    @DoNotStrip
    @Keep
    public static void addFrameProcessorPlugin(String name, PluginInitializer pluginInitializer) {
        assert !Plugins.containsKey(name) : "Tried to add a Frame Processor Plugin with a name that already exists! " +
                "Either choose unique names, or remove the unused plugin. Name: ";
        Plugins.put(name, pluginInitializer);
    }

    @DoNotStrip
    @Keep
    public static FrameProcessorPlugin getPlugin(String name, Map<String, Object> options) {
        PluginInitializer initializer = Plugins.get(name);
        if (initializer == null) {
            return null;
        }
        return initializer.initializePlugin(options);
    }

    public interface PluginInitializer {
        FrameProcessorPlugin initializePlugin(@Nullable Map<String, Object> options);
    }
}

File: package/ios/Frame Processor/FrameProcessorPluginRegistry.m
//
//  FrameProcessorPluginRegistry.m
//  VisionCamera
//
//  Created by Marc Rousavy on 24.03.21.
//  Copyright © 2021 mrousavy. All rights reserved.
//

#import "FrameProcessorPluginRegistry.h"
#import <Foundation/Foundation.h>

@implementation FrameProcessorPluginRegistry

+ (NSMutableDictionary<NSString*, PluginInitializerFunction>*)frameProcessorPlugins {
  static NSMutableDictionary<NSString*, PluginInitializerFunction>* plugins = nil;
  if (plugins == nil) {
    plugins = [[NSMutableDictionary alloc] init];
  }
  return plugins;
}

+ (void)addFrameProcessorPlugin:(NSString*)name
                withInitializer:(PluginInitializerFunction)pluginInitializer {
  BOOL alreadyExists =
      [[FrameProcessorPluginRegistry frameProcessorPlugins] valueForKey:name] != nil;
  NSAssert(!alreadyExists,
           @"Tried to add a Frame Processor Plugin with a name that already exists! Either choose "
           @"unique names, or "
           @"remove the unused plugin. Name: %@",
           name);

  [[FrameProcessorPluginRegistry frameProcessorPlugins] setValue:pluginInitializer forKey:name];
}

+ (FrameProcessorPlugin*)getPlugin:(NSString* _Nonnull)name
                       withOptions:(NSDictionary* _Nullable)options {
  PluginInitializerFunction initializer =
      [[FrameProcessorPluginRegistry frameProcessorPlugins] objectForKey:name];
  if (initializer == nil) {
    return nil;
  }

  return initializer(options);
}

@end
